examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,837 @@
+            self.owners.swap_remove(self.owner_index(&owner));
+            self.is_owner.remove(&owner);
+            self.requirement.set(requirement);
+            self.clean_owner_confirmations(&owner);
+            self.env().emit_event(OwnerRemoval { owner });
+        }
+
+        /// Replace an owner from the contract with a new one.
+        /// Panics if `old_owner` is no owner or if `new_owner` already is one.
+        /// Only callable by the wallet itself.
+        #[ink(message)]
+        fn replace_owner(&mut self, old_owner: AccountId, new_owner: AccountId) {
+            self.ensure_from_wallet();
+            self.ensure_owner(&old_owner);
+            self.ensure_no_owner(&new_owner);
+            self.owners
+                .replace(self.owner_index(&old_owner), || new_owner);
+            self.is_owner.remove(&old_owner);
+            self.is_owner.insert(new_owner, ());
+            self.clean_owner_confirmations(&old_owner);
+            self.env().emit_event(OwnerRemoval { owner: old_owner });
+            self.env().emit_event(OwnerAddition { owner: new_owner });
+        }
+
+        /// Change the requirement to a new value.
+        /// Only callable by the wallet itself.
+        #[ink(message)]
+        fn change_requirement(&mut self, new_requirement: u32) {
+            self.ensure_from_wallet();
+            self.ensure_requirement_is_valid(self.owners.len(), new_requirement);
+            self.requirement.set(new_requirement);
+            self.env().emit_event(RequirementChange { new_requirement });
+        }
+
+        /// Add a new transaction candiate to the contract.
+        /// This also confirms the transaction for the caller.
+        /// This can be called by any owner.
+        #[ink(message)]
+        fn submit_transaction(&mut self, transaction: Transaction) {
+            self.ensure_caller_is_owner();
+            let trans_id = self.transactions.put(transaction);
+            self.confirmation_count.insert(trans_id, 0);
+            self.env().emit_event(Submission {
+                transaction: trans_id,
+            });
+            self.confirm_by_caller(self.env().caller(), trans_id);
+        }
+
+        /// Remove a transaction from the contract.
+        /// Only callable by the wallet itself.
+        /// Panics if `trans_id` is no valid transaction id.
+        #[ink(message)]
+        fn cancel_transaction(&mut self, trans_id: TransactionId) {
+            self.ensure_from_wallet();
+            if self.take_transaction(trans_id).is_some() {
+                self.env().emit_event(Cancelation {
+                    transaction: trans_id,
+                });
+            }
+        }
+
+        /// Confirm a transaction for the sender that was submitted by any owner.
+        /// This can be called by any owner.
+        /// Panics if `trans_id` is no valid transaction id.
+        #[ink(message)]
+        fn confirm_transaction(&mut self, trans_id: TransactionId) {
```

Comment:
Maybe this could profit from a result that tells the caller if the transaction is ready for execution (has enough approvals). What do you think? This way you can more easily emulate the old-style of executing such a transaction right away. Or we could provide another method that does this job of the old semantics.

Reply:
This makes sense. Now that the confirmation check is cheap this might be viable to return something useful here.

Reply:
I added a `ConfirmationStatus`type that is returned from `submit_transaction` and `confirm_transaction`. It is also included in the confirmation event.